File: fogl/js/GlContextPeer.js
//
// Copyright (c) 2011, chunquedong
// Licensed under the Academic Free License version 3.0
//
// History:
//   2011-5-29  Jed Young  Creation
//


fan.fogl.GlContextPeer = fan.sys.Obj.$extend(fan.sys.Obj);
fan.fogl.GlContextPeer.prototype.$ctor = function(self) {}
fan.fogl.GlContextPeer.prototype.gl = null;

//////////////////////////////////////////////////////////////////////////
// common
//////////////////////////////////////////////////////////////////////////

fan.fogl.GlContextPeer.prototype.clearColor = function(self, r, g, b, a)
{
  this.gl.clearColor(r, g, b, a);
}

fan.fogl.GlContextPeer.prototype.enable = function(self, cap)
{
  this.gl.enable(cap.m_val);
}

fan.fogl.GlContextPeer.prototype.viewport = function(self, x, y, width, height)
{
  this.gl.viewport(x, y, width, height);
}

fan.fogl.GlContextPeer.prototype.clear = function(self, mask)
{
  this.gl.clear(mask.m_val);
}

fan.fogl.GlContextPeer.prototype.vertexAttribPointer = function(self, index, size, type, normalized, stride, offset)
{
  this.gl.vertexAttribPointer(index, size, type.m_val, normalized, stride, offset);
}

fan.fogl.GlContextPeer.prototype.drawArrays = function(self, mode, first, count)
{
  this.gl.drawArrays(mode.m_val, first, count);
}

//////////////////////////////////////////////////////////////////////////
// buffer
//////////////////////////////////////////////////////////////////////////

fan.fogl.GlContextPeer.prototype.createBuffer = function(self)
{
  var buf = fan.fogl.Buffer.make();
  buf.peer.setValue(this.gl.createBuffer());
  return buf;
}

fan.fogl.GlContextPeer.prototype.bindBuffer = function(self, target, buffer)
{
  this.gl.bindBuffer(target.m_val, buffer.peer.getValue());
}

fan.fogl.GlContextPeer.prototype.bufferData = function(self, target, data, usage)
{
  this.gl.bufferData(target.m_val, data.getData(), usage.m_val);
}

//////////////////////////////////////////////////////////////////////////
// shader
//////////////////////////////////////////////////////////////////////////

fan.fogl.GlContextPeer.prototype.createShader = function(self, type)
{
  var i = this.gl.createShader(type.m_val);
  var shader = fan.fogl.Shader.make();
  shader.peer.setValue(i);
  return shader;
}

fan.fogl.GlContextPeer.prototype.shaderSource = function(self, shader, source)
{
  this.gl.shaderSource(shader.peer.getValue(), source);
}

fan.fogl.GlContextPeer.prototype.compileShader = function(self, shader)
{
  this.gl.compileShader(shader.peer.getValue());
}

fan.fogl.GlContextPeer.prototype.getShaderParameter = function(self, shader, pname)
{
  return this.gl.getShaderParameter(shader.peer.getValue(), pname.m_val);
}

fan.fogl.GlContextPeer.prototype.getShaderInfoLog = function(self, shader)
{
  return this.gl.getShaderInfoLog(shader.peer.getValue());
}

fan.fogl.GlContextPeer.prototype.createProgram = function(self)
{
  var i = this.gl.createProgram();
  var program = fan.fogl.Program.make();
  program.peer.setValue(i);
  return program;
}

fan.fogl.GlContextPeer.prototype.attachShader = function(self, program, shader)
{
  this.gl.attachShader(program.peer.getValue(), shader.peer.getValue());
}

fan.fogl.GlContextPeer.prototype.linkProgram = function(self, program)
{
  this.gl.linkProgram(program.peer.getValue());
}

fan.fogl.GlContextPeer.prototype.getProgramParameter = function(self, program, pname)
{
  return this.gl.getProgramParameter(program.peer.getValue(), pname.m_val);
}

fan.fogl.GlContextPeer.prototype.validateProgram = function(self, program)
{
  this.gl.validateProgram(program.peer.getValue());
}

fan.fogl.GlContextPeer.prototype.useProgram = function(self, program)
{
  this.gl.useProgram(program.peer.getValue());
}

//////////////////////////////////////////////////////////////////////////
// uniform
//////////////////////////////////////////////////////////////////////////

fan.fogl.GlContextPeer.prototype.getUniformLocation = function(self, program, name)
{
  var i = this.gl.getUniformLocation(program.peer.getValue(), name);
  var location = fan.fogl.UniformLocation.make();
  location.peer.setValue(i);
  return location;
}

fan.fogl.GlContextPeer.prototype.uniformMatrix4fv = function(self, location, transpose, value)
{
  this.gl.uniformMatrix4fv(location.peer.getValue(), transpose, value.getData());
}

//////////////////////////////////////////////////////////////////////////
// vertexShader
//////////////////////////////////////////////////////////////////////////

fan.fogl.GlContextPeer.prototype.getAttribLocation = function(self, program, name)
{
  return this.gl.getAttribLocation(program.peer.getValue(), name);
}

fan.fogl.GlContextPeer.prototype.enableVertexAttribArray = function(self, index)
{
  this.gl.enableVertexAttribArray(index);
}

File: fogl/java/GlContextPeer.java
//
// Copyright (c) 2011, chunquedong
// Licensed under the Academic Free License version 3.0
//
// History:
//   2011-5-31  Jed Young  Creation
//

package fan.fogl;

import fan.sys.*;

import static org.lwjgl.opengl.GL11.*;
import static org.lwjgl.opengl.GL12.*;
import static org.lwjgl.opengl.GL13.*;
import static org.lwjgl.opengl.GL14.*;
import static org.lwjgl.opengl.GL15.*;
import static org.lwjgl.opengl.GL20.*;
import static org.lwjgl.opengl.GL21.*;
import static org.lwjgl.opengl.GL30.*;
import static org.lwjgl.opengl.GL31.*;

class GlContextPeer
{
  public static GlContextPeer make(GlContext self)
  {
    return new GlContextPeer();
  }

  public void clearColor(GlContext self, double r, double g, double b, double a)
  {
    glClearColor((float)r, (float)g, (float)b, (float)a);
  }

//////////////////////////////////////////////////////////////////////////
// common
//////////////////////////////////////////////////////////////////////////

  public void enable(GlContext self, GlEnum cap)
  {
    glEnable((int)cap.val);
  }

  public void viewport(GlContext self, long x, long y, long width, long height)
  {
    glViewport((int)x, (int)y, (int)width, (int)height);
  }

  public void clear(GlContext self, GlEnum mask)
  {
    glClear((int)mask.val);
  }

  public void vertexAttribPointer(GlContext self, long indx, long size, GlEnum type, boolean normalized, long stride, long offset)
  {
    glVertexAttribPointer((int)indx, (int)size, (int)type.val, normalized, (int)stride, offset);
  }

  public void drawArrays(GlContext self, GlEnum mode, long first, long count)
  {
    glDrawArrays((int)mode.val, (int)first, (int)count);
  }

//////////////////////////////////////////////////////////////////////////
// buffer
//////////////////////////////////////////////////////////////////////////

  public Buffer createBuffer(GlContext self)
  {
    int i = glGenBuffers();
    Buffer buf = Buffer.make();
    buf.peer.setValue(i);
    return buf;
  }

  public void bindBuffer(GlContext self, GlEnum target, Buffer buffer)
  {
    glBindBuffer((int)target.val, buffer.peer.getValue());
  }

  public void bufferData(GlContext self, GlEnum target, ArrayBuffer data, GlEnum usage)
  {
    java.nio.Buffer d = data.peer.getValue();
    if (d instanceof java.nio.FloatBuffer)
    {
      glBufferData((int)target.val, (java.nio.FloatBuffer)d, (int)usage.val);
    }
    else if (d instanceof java.nio.DoubleBuffer)
    {
      glBufferData((int)target.val, (java.nio.DoubleBuffer)d, (int)usage.val);
    }
    else if(d instanceof java.nio.IntBuffer)
    {
      glBufferData((int)target.val, (java.nio.IntBuffer)d, (int)usage.val);
    }
    else if(d instanceof java.nio.ShortBuffer)
    {
      glBufferData((int)target.val, (java.nio.ShortBuffer)d, (int)usage.val);
    }
    else if(d instanceof java.nio.ByteBuffer)
    {
      glBufferData((int)target.val, (java.nio.ByteBuffer)d, (int)usage.val);
    }
    else
    {
      throw UnsupportedErr.make("unsupported type");
    }
  }

//////////////////////////////////////////////////////////////////////////
// shader
//////////////////////////////////////////////////////////////////////////

  public Shader createShader(GlContext self, GlEnum type)
  {
    int i = glCreateShader((int)type.val);
    Shader shader = Shader.make();
    shader.peer.setValue(i);
    return shader;
  }

  public void shaderSource(GlContext self, Shader shader, String source)
  {
    glShaderSource(shader.peer.getValue(), source);
  }

  public void compileShader(GlContext self, Shader shader)
  {
    glCompileShader(shader.peer.getValue());
  }

  public boolean getShaderParameter(GlContext self, Shader shader, GlEnum pname)
  {
    int i = glGetShader(shader.peer.getValue(), (int)pname.val);
    return i != 0;
  }

  public String getShaderInfoLog(GlContext self, Shader shader)
  {
    return glGetShaderInfoLog(shader.peer.getValue(), 1024);
  }


  public Program createProgram(GlContext self)
  {
    int i = glCreateProgram();
    Program p = Program.make();
    p.peer.setValue(i);
    return p;
  }

  public void attachShader(GlContext self, Program program, Shader shader)
  {
    glAttachShader(program.peer.getValue(), shader.peer.getValue());
  }

  public void linkProgram(GlContext self, Program program)
  {
    glLinkProgram(program.peer.getValue());
  }

  public boolean getProgramParameter(GlContext self, Program program, GlEnum pname)
  {
    int i = glGetProgram(program.peer.getValue(), (int)pname.val);
    return i != 0;
  }

  public void validateProgram(GlContext self, Program program)
  {
    glValidateProgram(program.peer.getValue());
  }

  public void useProgram(GlContext self, Program program)
  {
    glUseProgram(program.peer.getValue());
  }

//////////////////////////////////////////////////////////////////////////
// uniform
//////////////////////////////////////////////////////////////////////////

  public UniformLocation getUniformLocation(GlContext self, Program program, String name)
  {
    int i = glGetUniformLocation(program.peer.getValue(), name);
    UniformLocation location = UniformLocation.make();
    location.peer.setValue(i);
    return location;
  }

  public void uniformMatrix4fv(GlContext self, UniformLocation location, boolean transpose, ArrayBuffer value)
  {
    java.nio.Buffer d = value.peer.getValue();
    if (d instanceof java.nio.FloatBuffer)
    {
      glUniformMatrix4(location.peer.getValue(), transpose, (java.nio.FloatBuffer)d);
    }
    else
    {
      throw UnsupportedErr.make("unsupported type");
    }
  }

//////////////////////////////////////////////////////////////////////////
// vertexShader
//////////////////////////////////////////////////////////////////////////

  public long getAttribLocation(GlContext self, Program program, String name)
  {
    return glGetAttribLocation(program.peer.getValue(), name);
  }

  public void enableVertexAttribArray(GlContext self, long index)
  {
    glEnableVertexAttribArray((int)index);
  }
}